### ISSUE TITLE
Unable to publish with additional custom attributes

### ISSUE THREAD

**Issue by joseph-lum** [mia-0032/fluent-plugin-gcloud-pubsub-custom#56]
Is there a way to publish to pubsub with custom attributes like in the pubsub documentation: https://cloud.google.com/pubsub/docs/reference/rest/v1/PubsubMessage

I tried doing logger.emit(message, attributes), but it doesn't seem to work. Is it possible to amend the library with the following documentation to add functionality for publishing with custom attributes like: https://cloud.google.com/pubsub/docs/publisher#pubsub-publish-custom-attributes-ruby

**Comment by mia-0032** [comment]
Hi,

> Is there a way to publish to pubsub with custom attributes like in the pubsub documentation: 

This plugin don't have the feature publishing messages with attributes yet.

> I tried doing logger.emit(message, attributes), but it doesn't seem to work.

Sorry, I couldn't understand this. Do you want to use Fluentd Logger?
Even if this plugin have that feature, you must write Fluentd configuration file such as written in README.

> Is it possible to amend the library with the following documentation to add functionality for publishing with custom attributes like: 

I think that it is possible. But its feature is able to use as plugin not library. As I mentioned before
, you have to write Fluentd configuration file.

**Comment by joseph-lum** [comment]
Hi, thanks for your reply. Yes, I am using it as a plugin with the Fluentd logger. Currently, I made it work by modifying /fluent-plugin-gcloud-pubsub-custom-1.2.0/lib/fluent/plugin/gcloud_pubsub/client.rb with the following lines:

```
def publish(topic_name, messages)
  topic(topic_name).publish do |batch|
    messages.each do |m|
      datahash = JSON.parse(m)
      batch.publish datahash["data"].to_json, datahash["attributes"]  # custom attributes published here
    end
  end
rescue Google::Cloud::UnavailableError, Google::Cloud::DeadlineExceededError, Google::Cloud::InternalError => ex
  raise RetryableError.new "Google api returns error: #{ex.class.to_s} message: #{ex.to_s}"
end
```

I guess it would be great if there's a possibility to publish custom attributes via configuration settings in the Fluentd configuration file, but otherwise I would have to live with this small manual tweak.

**Comment by mia-0032** [comment]
Hi, thank you for your reply.
I understood the feature that you want. I'll develop it later. I'll reply to this issue when I complete to develop.

**Comment by joseph-lum** [comment]
Thanks @mia-0032!! That would be really great :)

**Comment by mia-0032** [comment]
I released v1.3.0.
If the version have any problem, please open new issue.
I'll close this issue.
